Rlabkey/R/labkey.experiment.R

```diff
@@ -0,0 +1,173 @@
+
+    material <- config
+
+    if (!is.null(sampleSetId) || !is.null(sampleSetName))
+    {
+        sampleSet <- list()
+        if (!is.null(sampleSetId))
+            sampleSet$id = sampleSetId
+        if (!is.null(sampleSetName))
+            sampleSet$name = sampleSetName
+
+        material$sampleSet = sampleSet
+    }
+    return (material)
+}
+
+## Create an ExpRun object
+##
+labkey.experiment.createRun <- function(config, dataInputs = NULL, dataOutputs = NULL, dataRows = NULL, materialInputs = NULL, materialOutputs = NULL)
+{
+    ## check required parameters
+    if (missing(config))
+        stop (paste("A list of ExpObject config params must be specified in the config parameter."))
+
+    run <- config
+
+    if (!is.null(dataInputs))
+    {
+        if (!is.list(dataInputs))
+            stop (paste("dataInputs must be a list of data objects, see labkey.experiment.createData."))
+
+        ## ensure dataInputs is serialized as an array of objects
+        run$dataInputs = ensureNestedList(dataInputs)
+    }
+
+    if (!is.null(dataOutputs))
+    {
+        if (!is.list(dataOutputs))
+            stop (paste("dataOutputs must be a list of data objects, see labkey.experiment.createData."))
+
+        run$dataOutputs = ensureNestedList(dataOutputs)
+    }
+
+    if (!is.null(materialInputs))
+    {
+        if (!is.list(materialInputs))
+            stop (paste("materialInputs must be a list of material objects, see labkey.experiment.createMaterial."))
+
+        run$materialInputs = ensureNestedList(materialInputs)
+    }
+
+    if (!is.null(materialOutputs))
+    {
+        if (!is.list(materialOutputs))
+            stop (paste("materialOutputs must be a list of material objects, see labkey.experiment.createMaterial."))
+
+        run$materialOutputs = ensureNestedList(materialOutputs)
+    }
+
+    if (!is.null(dataRows))
+    {
+        if (!is.data.frame(dataRows))
+            stop (paste("dataRows must be a data frame."))
+
+    	## build Assay object tree based on R lists
+    	nrows <- nrow(dataRows)
+    	ncols <- ncol(dataRows)
+    	cnames <- colnames(dataRows)
+    	rowsVector <- vector(mode="list", length=nrows)
+    	for (j in 1:nrows)
+    	{
+    		cvalues <- as.list(dataRows[j,])
+    		names(cvalues) <- cnames
+    		rowsVector[[j]] <- cvalues
+    	}
+
+        run$dataRows <- rowsVector
+    }
+    return (run)
+}
+
+## Helper to ensure the passed object is serialized to JSON as an array of objects
+##
+ensureNestedList <- function(data)
+{
+    if (is.null(names(data)))
+        data
+    else
+        data <- list(data)
+
+    return (data)
+}
+
+labkey.experiment.saveBatch <- function(baseUrl=NULL, folderPath, assayConfig = NULL, protocolName = NULL, runList)
+{
+    baseUrl=labkey.getBaseUrl(baseUrl)
+
+    ## check required parameters
+    if (missing(baseUrl) || is.null(baseUrl) || missing(folderPath) || missing(runList))
```

Comment:
baseUrl shouldn't be required, it could have been set previously and available at labkey.getBaseUrl. If baseUrl is required for saveBatch, than need to change function signature and remove '=NULL'.